content/blog/2020-03-06-wasm/index.md

```diff
@@ -0,0 +1,229 @@
+i.e., platform independent
+- it is programming model independent
+- it is hardware independent
+
+What this means is you could write a web application in C, compile it to 
+WebAssembly and use it on any browser on any hardware. This gives much more 
+performance in general than a JavaScript program, which has type 
+classification overheads and overheads at parsing and verification.
+Existing technologies also has performance implications based on which browser 
+you target (different browsers used different techniques to optimize applications on
+browsers).
+
+## Is WebAssembly just C on browsers?
+WebAssembly does brings the performance of pre-compiled languages to web browsers. 
+But it does so side-by-side with traditional JavaScript running on the same engines.
+This allows browsers to have best of both worlds, drive the Ferrari on the race track--- 
+run performant WebAssembly when it needs performance--- or drive the Prius to the 
+grocery store--- quickly getting a simple web application up and running.
+
+With WebAssembly, you could use JavaScript for fast development, but also 
+use C where you need performance. You could use C (theoretically supports other languages too) where static typing is
+useful, but JavaScript where dynamic typing is a necessity for productivity.
+And you could combine these modules, leveraging interoperability to balance productivity and performance,
+as both these styles can now be executed in the same compiler flow.
+So WebAssembly is more than just C on browsers, it's a carefully coordinated, 
+well-designed, massive engineering effort to reconcile two worlds at odds.
+
+## How does it differ from existing tools?
+So WebAssembly is portable, but doesn't JVM (Java virtual machine) already 
+offer portability? Even though the overall goal has some similarities, Java 
+cannot be run on a browser without plugins and has doesn't have the language
+flexibilty WebAssembly provides. This is due to Java's portability coming 
+from the ability to run Java programs in different platforms, and therefore not
+designed to run multiple programs in a browser. Java also doesn't solve the issues
+JavaScript poses through interpreter based execution, dynamic typing and
+garbage collection. Moreover, validation and memory safety makes 
+WebAssembly more secure from vulnerabilities. Java was not designed with 
+formalization in mind, and therefore is complex to validate. Java applets
+have strict security restrictions as a work around for many implementation bugs.
+Structured control flow and single pass compilation and validation places WebAssembly as a 
+mature option compared to existing bytecode formats such as JVM and CIL 
+(Common Intermediate Language) which contain irreducible loops and unbalanced locking 
+features that are hard to be handled with JIT compilers and are relegated to 
+the interpreter. And these design decisions has made WebAssembly a better 
```

Comment:
It is not true that irreducible loops must be relegated to the interpreter. And Java's mutexes are lexically scoped by default (via `synchronized`) so I don't think "unbalanced locking" is a problem there?

Reply:
I guess I don’t quite follow the last para in section 8 in https://people.mpi-sws.org/%7Erossberg/papers/Haas,%20Rossberg,%20Schuff,%20Titzer,%20Gohman,%20Wagner,%20Zakai,%20Bastien,%20Holman%20-%20Bringing%20the%20Web%20up%20to%20Speed%20with%20WebAssembly.pdf
I was trying to express that claim.

Reply:
It also has a footnote in section 7 which seems to contradict this paragraph, and agree with your assessment above.

Reply:
I'm going to merge the PR now so this is resolved, but if you want, consider coming back to this assertion. I still think it's not technically accurate to say JVM & CIL "contain irreducible loops and unbalanced locking 
 features that are hard to be handled with JIT compilers and are relegated to 
 the interpreter."